### ISSUE TITLE
[Rails 5.1] Fix dirty.rb to use the new API.  *_changed?, *_was...

### ISSUE THREAD

**Issue by Erowlin** [mbleigh/acts-as-taggable-on#836]
This https://github.com/mbleigh/acts-as-taggable-on/blob/8e64c3d4a81cfbb8af621228eae36a65c1f94501/lib/acts_as_taggable_on/taggable/dirty.rb should be rewritten to use the new Rails 5.1 API Dirty. 

Here is the list of changes introduced in Rails 5.1: 
https://github.com/rails/rails/pull/25337#issuecomment-225166796

Relevant ones for acts-as-taggable:
- `def #{tag_type}_list_changed?` -> `def will_save_change_to_#{tag_type}_list?`
- `def #{tag_type}_list_was` -> `def #{tag_type}_list_in_database_was`
- `def #{tag_type}_list_change` -> `def #{tag_type}_list_ change_to_be_saved`
- `def #{tag_type}_list_changes` -> **Not sure about that one.** 

Otherwise, it's gonna trigger **a lot** of deprecation warnings : 
```
DEPRECATION WARNING: The behavior of `attribute_changed?` inside of after callbacks will be changing in the next version of Rails. The new return value will reflect the behavior of calling the method after `save` returned (e.g. the opposite of what it returns now). To maintain the current behavior, use `saved_change_to_attribute?` instead.
```

**Comment by Erowlin** [comment]
Actually it's broken. 
```ruby
user.my_tag_list
#=> ['hello'] 
user.my_tag_list = 'world'
user.my_tag_list_changed?
#=> false
```
While it should returns true.

**Comment by haraldkrischner** [comment]
I can confirm this. Have you thought about a workaround?

**Comment by Erowlin** [comment]
Yes we found a dirty hack which is basically in our controller

```ruby
previous_tag_list = user.my_tag_list.to_set
user.assign_attributes(params)
do_something if previous_tag_list != user.my_tag_list.to_set
```

Without this hack, it would look like that: 
```ruby
user.assign_attributes(params)
do_something if user.will_save_change_to_my_tag_list
```

**Comment by haraldkrischner** [comment]
Thanks for your quick answer, will have a look into it

**Comment by Erowlin** [comment]
Another very very important note about that, which I did not mentioned in my original comment: actually the original model is not dirtified also, while it should be: 
```ruby

user. my_tag_list = ['123']
# Using the old API for the example, note that `changed?` has been replaced by `has_changes_to_save`
user.my_tag_list.changed?
#=> false
``` 

while it should returns true. 
This is highly critical, maybe more important than the initial fix. 

Thanks for your answer @haraldkrischner !

**Comment by Erowlin** [comment]
> Sadly this does not look like a solution for a call inside an after_save callback.

You're right, we don't have such use case in callbacks in our application.

**Comment by dhruvbhatia** [comment]
Rails noob here, but I think this is breaking compatibility with [the paper trail gem](https://github.com/airblade/paper_trail/), as it no longer tracks changes to tag lists in Rails 5 despite it [previously working in Rails 4](https://github.com/airblade/paper_trail/issues/224).

**Comment by djpate** [comment]
Is anyone looking into this? I'm about to take a stab at it but I want to make sure I'm not duplicating work.

**Comment by Erowlin** [comment]
Not looking into it and not sure why it's been closed.

**Comment by haraldkrischner** [comment]
#861

**Comment by garymh** [comment]
Getting the same problem as @Erowlin... `model_name.changed?` is returning false for me. Sadly, #861 doesn't seem to fix it. Anyone figure out a solution?

**Comment by haraldkrischner** [comment]
Try #862 and tell me if you are lucky.

**Comment by garymh** [comment]
@haraldkrischner Nope :( I actually can't even get my tags to save with #862.

**Comment by Fodoj** [comment]
I've tried to make gem work on Rails 5.2 and Rails 5.1, including correct behaviour for Dirty. Tests pass and my application works well, but I don't have a confidence in the result. I was digging into source of Rails for couple of hours and might not see it clear enough. Would anyone be willing to give it a try? My current code is here: https://github.com/mbleigh/acts-as-taggable-on/pull/887

**Comment by Erowlin** [comment]
Hey @Fodoj thanks for taking care of it. 
I reviewed your PR, I am not confident enough with the Acts-As-Taggable code to say that everything is fine ;).

**Comment by seuros** [comment]
Fixed in 6.0.0
